src/main/java/org/mongojack/JacksonMongoCollection.java

```diff
@@ -0,0 +1,1246 @@
+/*
+ * Copyright 2011 VZ Netzwerke Ltd
+ * Copyright 2014 devbliss GmbH
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.mongojack;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.bson.BsonObjectId;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+import org.mongojack.internal.MongoJackModule;
+import org.mongojack.internal.object.document.DocumentObjectGenerator;
+import org.mongojack.internal.object.document.DocumentObjectTraversingParser;
+import org.mongojack.internal.query.QueryCondition;
+import org.mongojack.internal.util.DocumentSerializationUtils;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.mongodb.MongoBulkWriteException;
+import com.mongodb.MongoException;
+import com.mongodb.MongoWriteConcernException;
+import com.mongodb.MongoWriteException;
+import com.mongodb.ReadPreference;
+import com.mongodb.WriteConcern;
+import com.mongodb.client.AggregateIterable;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MapReduceIterable;
+import com.mongodb.client.model.FindOneAndUpdateOptions;
+import com.mongodb.client.model.IndexOptions;
+import com.mongodb.client.model.ReturnDocument;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.result.DeleteResult;
+import com.mongodb.client.result.UpdateResult;
+
+/**
+ * A DBCollection that marshals/demarshals objects to/from Jackson annotated
+ * classes. It provides a very thin wrapper over an existing MongoCollection.
+ * 
+ * A JacksonMongoCollection is threadsafe, with a few caveats:
+ * 
+ * If you pass your own ObjectMapper to it, it is not thread safe to reconfigure
+ * that ObjectMapper at all after creating it. The setWritePreference and a few
+ * other methods on JacksonMongoCollection should not be called from multiple
+ * threads
+ * 
+ * @author James Roper
+ * @since 1.0
+ */
+public class JacksonMongoCollection<T> {
+
+    private static final ObjectMapper DEFAULT_OBJECT_MAPPER = MongoJackModule
+            .configure(new ObjectMapper());
+
+    private com.mongodb.client.MongoCollection<T> mongoCollection;
```

Comment:
can this be final? do we need the fully specified package name or can we just import it instead?

Reply:
Can't be final if we want to let users set WriteConcern or ReadConcern, unless we do the same thing as the Mongo driver here and return a new instance if the JacksonMongoCollection.

Fully specified package name is because there is already a class that has the same name in the package as the JacksonMongoCollection.